### DISCUSSION TITLE
Pod restart/delete will result in node loss in graph

### DISCUSSION THREAD

**Discussion by garyyang85** [kiali/kiali#4175] [General]
It is the bookinfo example. When all the pods are running, access the productpage, will generate a wonderful graph in kiali. But when some pods are delete/restart, e.g. productpage and details.  the two related notes in graph will loss. Access the productpage again, the lost node will be back.
Work as design? Thanks.

**Comment by xeviknal** [comment]
hi @garyyang85,

I've converted this issue into a discussion. This way we can understand your scenario first and see what's wrong. If we can detect the bug then we move back this to an issue. Anyway, thank you for coming and starting a conversation.

The kiali graph is based in the available telemetry (prometheus metrics) during the selected window of time. It means that if there is no traffic the graph will be empty.

If you want to keep seeing the unused services/workloads, you can enable the following options. You won't see the traffic flowing between them though.

![Screenshot of Kiali Console (12)](https://user-images.githubusercontent.com/613814/125595850-7cbdd689-8247-486b-8720-bc2f087d26a5.png)

**Reply by xeviknal** [reply]
Is that information useful for you? Does it help?

I am also thinking that if you change the time window which kiali shows the graph/health, you might be able to see the traffic and nodes. For example, if you restarted one pod (or the pod had a problem) 5 minutes ago and your time window is 1minute, kiali won't show any traffic. However, kiali will show the traffic registered when everything went alright if you put the time window to something greater than 5 minutes.

![Screenshot of Kiali Console (13)](https://user-images.githubusercontent.com/613814/125597318-410bc5e4-5072-468e-8787-f362eb05990a.png)

**Reply by garyyang85** [reply]
Got it. Thanks for your quick reply. But seems my version 1.24 do not have such ability. :(

**Reply by xeviknal** [reply]
I am pretty sure there is a similar functionality. It might have a different naming and slightly different behavior. If you can paste the display dropdown, I can point it out quickly. I don't have a 1.24 ready to deploy.

**Reply by garyyang85** [reply]
the time window is there, I can change it to get old nodes :)

**Reply by garyyang85** [reply]
<img width="1919" alt="image" src="https://user-images.githubusercontent.com/20335728/125641891-38c0a912-7376-432b-abaf-da0609e49f42.png">

**Reply by xeviknal** [reply]
Good! And you can also use the option `Unused nodes` from the `Display` dropdown to show nodes that don't have traffic in the present range window (If I am not wrong).

**Reply by garyyang85** [reply]
@xeviknal here is my graph...  I am using openshift route(not istio ingress gateway) to access productpage. Quite confused with unknown to v1/v2/v3 and PassthroughCluster

**Reply by garyyang85** [reply]
<img width="1916" alt="image" src="https://user-images.githubusercontent.com/20335728/125642760-0e0e0355-4c7e-4561-875d-0d13925b6825.png">

**Reply by xeviknal** [reply]
Have you had the chance to read our FAQ website? https://kiali.io/documentation/latest/faq/#many-unknown

Guessing from the graph and the fact that you are using openshift with routes, I believe that you have placed a route to the reviews service, right? So if you have traffic from `unknown` to `reviews v1/2/3` it means that there is traffic from outside of the mesh directly to one service without going through any gateway. Istio is not able to identify traffic from non-mesh sources.

**Reply by xeviknal** [reply]
Let me add a couple of things in here:
1) We have a repo full of demo applications for you to play around with istio: https://github.com/kiali/demos
2) There is a hack folder (full of tooling to support development) where you can deploy the bookinfo with a traffic generator. This generator is useful to see the graph always with traffic and play around with istio configs (dr, vs, sidecars, and so on): https://github.com/kiali/kiali/blob/master/hack/istio/install-bookinfo-demo.sh

**Reply by garyyang85** [reply]
Many thanks for your kindly reply. I am sure there is only one route for productpage service, no route for other service in this namespace. We do not use the istio ingress gateway for some internal reason. But we really like the wonderful features brought by kiali. I have read the faq section for "unknown nodes" and "passthgoughcluster" before, but seems still can not explain the behavior in my test. Anyway I would like to try the demos to see if any inspiration can be found there.

**Reply by jshaughn** [reply]
I think you are hitting the issue descibed in this [Istio Issue](https://github.com/istio/istio/issues/24379#issuecomment-669921820).    I think this was resolved in Istio 1.8, are you running an earlier version?  The comment I linked has a couple of workarounds.
